core/src/main/java/org/apache/iceberg/ContentFileParser.java

```diff
@@ -301,4 +291,41 @@ private static Metrics metricsFromJson(JsonNode jsonNode) {
         lowerBounds,
         upperBounds);
   }
+
+  private static void partitionToJson(
+      Types.StructType partitionType, StructLike partitionData, JsonGenerator generator)
+      throws IOException {
+    generator.writeStartArray();
+    List<Types.NestedField> fields = partitionType.fields();
+    for (int pos = 0; pos < fields.size(); ++pos) {
+      Types.NestedField field = fields.get(pos);
+      Object partitionValue = partitionData.get(pos, Object.class);
+      SingleValueParser.toJson(field.type(), partitionValue, generator);
+    }
+    generator.writeEndArray();
+  }
+
+  private static PartitionData partitionFromJson(
+      Types.StructType partitionType, JsonNode partitionNode) {
+    Preconditions.checkArgument(
+        partitionNode.isArray(),
+        "Invalid partition data for content file: non-array (%s)",
+        partitionNode);
+
+    List<Types.NestedField> fields = partitionType.fields();
+    Preconditions.checkArgument(
+        partitionNode.size() == fields.size(),
+        "Invalid partition data size: expected = %s, actual = %s",
+        fields.size(),
+        partitionNode.size());
+
+    PartitionData partitionData = new PartitionData(partitionType);
+    for (int pos = 0; pos < fields.size(); ++pos) {
+      Types.NestedField field = fields.get(pos);
+      Object partitionValue = SingleValueParser.fromJson(field.type(), partitionNode.get(pos));
+      partitionData.set(pos, partitionValue);
+    }
```

Comment:
can we add an UT where there are > 1 partitions and each partition with a different type, to check the order ?